### PR TITLE
Updating Generate Schedule

### PR DESCRIPTION
**Pull request by aqeelmozumder**
Refactored the code as the resolver index is getting big and messy.

**These are the steps that is going on in `generateSchedule` Mutation:**

- Checks if the `input.courses `are for Fall, Spring or Summer
- Gets all users for professors
- Connect to Algorithm 2 and sends courses in the post request which gives back new capacity for the courses. => (src/resolverUtils.ts) 
- Connect to Algorithm 1 and sends Fall, Spring or Summer `coursesToSchedule` and users for `professors`, which gives back the scheduled data (right now giving null) =>  (src/resolverUtils.ts)
- Creates Schedule and upserts the courses as before, just refactored
- Seed Professor and Preferences in DB
- Ensure Algorithm 1 Response is giving data
- Hook up Algorithm 1 Capacity with Algorithm 2 Capacity

### CODE REVIEW

File: src/prisma/course.ts [hunk 1]
```diff
@@ -17,38 +18,112 @@ async function findCourseSection(courseterm: Term) {
...
+
+    if (term === Term.FALL) {
+      startTime.setFullYear(year, 10, 1);
+      endTime.setFullYear(year, 11, 1);
+    } else if (term === Term.SPRING) {
+      startTime.setFullYear(year, 0, 1);
+      endTime.setFullYear(year, 3, 1);
+    } else if (term === Term.SUMMER) {
+      startTime.setFullYear(year, 4, 1);
+      endTime.setFullYear(year, 9, 1);
```
Comment (MustafaWasif) [inline src/prisma/course.ts:77]: endTime for Summer should be 7

File: src/prisma/course.ts [hunk 2]
```diff
@@ -17,38 +18,112 @@ async function findCourseSection(courseterm: Term) {
...
+        start: getDateTime(assignment.beginTime, assignment.startDate),
+        end: getDateTime(assignment.endTime, assignment.endDate),
+      };
+    }
+
+    const startTime = getTime(assignment.beginTime);
+    const endTime = getTime(assignment.endTime);
+
+    if (term === Term.FALL) {
+      startTime.setFullYear(year, 10, 1);
```
Comment (MustafaWasif) [inline src/prisma/course.ts:70]: startTime 8 instead of 10 for Fall term

File: src/prisma/course.ts (hunk 1 above)
Comment (aomi) [reply src/prisma/course.ts:77]: But yeah you're right about it being off. Fixing

File: src/prisma/course.ts [hunk 3]
```diff
@@ -14,3 +17,148 @@ async function findCourseSection(courseterm: Term) {
...
+  };
+};
+
+const getMinDate = (dates: Date[]): Date =>
+  new Date(Math.min(...dates.map((d) => d.getTime())));
+
+const getMaxDate = (dates: Date[]): Date =>
+  new Date(Math.max(...dates.map((d) => d.getTime())));
+
+async function upsertCourses(
```
Comment (LacroixJ) [inline src/prisma/course.ts:105]: minor nitpick
```suggestion
async function upsertCourse(
```

File: src/resolvers/resolverUtils.ts [hunk 4]
```diff
@@ -170,3 +183,101 @@ async function getSchedule(year: number): Promise<Schedule | null> {
...
+        return (
+          course.subject === input.subject &&
+          course.code === input.code &&
+          course.semester === term
+        );
+      })?.capacity ?? 0,
+    // default capacity to 0 if not found
+    courseTitle: 'testing',
+    // TODO: figure out best value here
+    sequenceNumber: 'testing',
```
Comment (LacroixJ) [inline src/resolvers/resolverUtils.ts:246]: sequence number is an oddly named variable but from how I've seen alg1 use it, it is supposed to be the section name(eg. A01).

For course title, I can see how that might be an issue, but also algorithm 1 really should not need us to submit this. If for some reason it becomes necessary and is not easily accessible, I might be able to store it in the same json file as the stream sequences. 

No changes needed here, just commenting.

File: src/prisma/course.ts (hunk 3 above)
Comment (LacroixJ) [reply src/prisma/course.ts:105]: I don't really care if this gets fixed.

**Review by LacroixJ — APPROVED** [review]

**Pull request merged**